dimod/reference/samplers/exact_solver.py

```diff
@@ -177,28 +176,93 @@ def __init__(self):
         self.properties = {}
         self.parameters = {}
 
-    def sample_dqm(self, dqm, **kwargs):
+    def sample_dqm(self, dqm, **kwargs) -> SampleSet:
         """Sample from a discrete quadratic model.
 
         Args:
-            dqm (:obj:`~dimod.DiscreteQuadraticModel`):
+            dqm (:class:`~dimod.DiscreteQuadraticModel`):
                 Discrete quadratic model to be sampled from.
 
         Returns:
-            :obj:`~dimod.SampleSet`
+            :class:`~dimod.SampleSet`
 
         """
         Sampler.remove_unknown_kwargs(self, **kwargs)
         
-        n = dqm.num_variables()
-        if n == 0:
+        if not dqm.num_variables():
             return SampleSet.from_samples([], 'DISCRETE', energy=[])
 
-        possible_samples = as_samples((_all_cases_dqm(dqm), list(dqm.variables)))
-        energies = dqm.energies(possible_samples)
+        cases = _all_cases_dqm(dqm)
+        energies = dqm.energies(cases)
+        
+        return SampleSet.from_samples(cases, 'DISCRETE', energies)
+    
+    
+class ExactCQMSolver():
+    """A simple exact solver for testing and debugging code using your local CPU.
+
+    Notes:
+        This solver calculates the energy for every possible
+        combination of variable cases. It becomes slow very quickly
+        
+    Examples:
+        This example solves a CQM with 3 variables and 1 constraint.
+        
+        >>> from dimod import ConstrainedQuadraticModel, Binary
+        >>> cqm = ConstrainedQuadraticModel()
+        >>> x, y, z = Binary('x'), Binary('y'), Binary('z')
+        >>> cqm.set_objective(x*y + 2*y*z)
+        >>> cqm.add_constraint(x*y == 1)      # doctest: +SKIP
+        >>> sampleset = ExactCQMSolver().sample_cqm(cqm)      # doctest: +SKIP
+        >>> print(sampleset)      # doctest: +SKIP
+          x y z energy num_oc. is_sat. is_fea.
+        0 0 0 0    0.0       1 arra...   False
+        1 0 1 0    0.0       1 arra...   False
+        2 1 0 0    0.0       1 arra...   False
+        4 0 0 1    0.0       1 arra...   False
+        6 1 0 1    0.0       1 arra...   False
+        3 1 1 0    1.0       1 arra...    True
+        5 0 1 1    2.0       1 arra...   False
+        7 1 1 1    3.0       1 arra...    True
+        ['INTEGER', 8 rows, 8 samples, 3 variables]
+        
+    """
+    properties = None
```

Comment:
These can be removed. We're not inheriting from `Sampler` so you don't need these class attributes to satisfy the ABC.

Reply:
So something like
```def __init__(self):
        pass```
and removing ```Sampler.remove_unknown_kwargs(self, **kwargs)``` would be fine?

Reply:
No, leave the `__init__` the way it is, just remove the
```python
properties = None
parameters = None
```